File: 251-flatten-2d-vector/251-flatten-2d-vector.py
class Vector2D:

    def __init__(self, vec: List[List[int]]):
        self.vec = [lst for lst in vec if lst]
        self.lst_no = 0
        self.pos = 0

    def next(self) -> int:
        v = self.vec[self.lst_no][self.pos]
        if self.pos + 1 == len(self.vec[self.lst_no]):
            self.lst_no += 1
            self.pos = 0
        else:
            self.pos += 1
        return v

    def hasNext(self) -> bool:
        return self.lst_no < len(self.vec) and self.pos < len(self.vec[self.lst_no])


# Your Vector2D object will be instantiated and called as such:
# obj = Vector2D(vec)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: 251-flatten-2d-vector/251-flatten-2d-vector.java
class Vector2D {
    private int[][] vec;
    private int v;
    private int i;
    
    public Vector2D(int[][] vec) {
        this.vec = vec;
    }
    
    public int next() {
        int ans = this.vec[this.v][this.i];
        if(this.i == this.vec[this.v].length - 1){
            this.v++;
            this.i = 0;
        } else {
            this.i++;
        }
        return ans;
    }
    
    public boolean hasNext() {
        while(this.v < this.vec.length && this.vec[this.v].length == 0){
            this.v++;
        }
        
        if(this.v < this.vec.length && this.i < this.vec[this.v].length){
            return true;
        } else {
            return false;
        }
    }
}

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D obj = new Vector2D(vec);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */